### PR TITLE
fix stderr output in python channels

### PR DESCRIPTION
**Pull request by timwr**
~~This is my attempt to fix the cmd_exec tests on python. They are currently intermittently failing on 77c9f6c86808e472798c52cbe51b9fb0ecca130f, and failing only the stderr tests on 0e2b68a9feda034fbec9ebc65fc4581390f89912~~

This change fixes the stderr output on python channels, which seems to be required after https://github.com/rapid7/metasploit-payloads/pull/510

Before and after this change, there still appears to be some channels left open, as `meterpreter > exit` gives lots of the following output:
```
Traceback (most recent call last):
  File "<string>", line 982, in send_packet
  File "<string>", line 1181, in _send_packet
error: [Errno 32] Broken pipe
Traceback (most recent call last):
  File "<string>", line 982, in send_packet
  File "<string>", line 1181, in _send_packet
error: [Errno 32] Broken pipe
```
The Traceback appears once for each stdapi_process_execute that was ran as part of the tests, as stdapi_process_close does not seem to properly close/cleanup the channel.

# Verification
- [x] Start msfconsole
- [x] Get a python/meterpreter/reverse_tcp session (I used localhost and python3)
- [x] Run the cmd_exec tests:
```
loadpath test/modules
use post/test/cmd_exec
set SESSION -1
set VERBOSE true
run
```
**** Before
```
msf6 post(test/cmd_exec) > run

[*] Running against session 6
[*] Session type is meterpreter and platform is linux
[*] Starting cmd_exec tests
[+] should return the result of echo
[+] should return the full response after sleeping
[+] should return the full response after sleeping
[*] Starting cmd_exec quote tests
[+] should return the result of echo with single quotes
[+] should return the result of echo with double quotes
[*] Starting cmd_exec stderr tests
[-] FAILED: should return the stderr output
[*] Testing complete in 2.802766206
[-] Passed: 6; Failed: 1
[*] Post module execution completed
msf6 post(test/cmd_exec) > irb
[*] Starting IRB shell...
[*] You are in post/test/cmd_exec

>> cmd_exec('invalidcommand')
=> ""

```

**** After
```
msf6 post(test/cmd_exec) > run

[*] Running against session -1
[*] Session type is meterpreter and platform is linux
[*] Starting cmd_exec tests
[+] should return the result of echo
[+] should return the full response after sleeping
[+] should return the full response after sleeping
[+] should return the result of echo 10 times
[*] Finished cmd_exec tests
[*] Starting cmd_exec quote tests
[+] should return the result of echo with single quotes
[+] should return the result of echo with double quotes
[*] Starting cmd_exec stderr tests
[+] should return the stderr output
[*] Testing complete in 4.801493159
[*] Passed: 7; Failed: 0
[*] Post module execution completed
msf6 post(test/cmd_exec) > irb
[*] Starting IRB shell...
[*] You are in post/test/cmd_exec

>> cmd_exec('invalidcommand')
=> ": 1: invalidcommand: not found"
```

### CODE REVIEW

**Comment by timwr** [comment]
Apologies for the confusion. There are lot of moving parts, and the bug I'm seeing is quite hard to reproduce (perhaps some kind of race condition). For me it occurs after running the unit tests, on a python/meterpreter/reverse_tcp session, with python3 on localhost, about 10 times.
So while this change seems to be required to fix the stderr output after https://github.com/rapid7/metasploit-payloads/pull/510, I don't think it addresses the intermittent failure described above.

Additionally the error output:
```
Traceback (most recent call last):
  File "<string>", line 982, in send_packet
  File "<string>", line 1181, in _send_packet
error: [Errno 32] Broken pipe
Traceback (most recent call last):
  File "<string>", line 982, in send_packet
  File "<string>", line 1181, in _send_packet
error: [Errno 32] Broken pipe
```
appears both before and after this change.
I'll update the title to reflect this and mark this as ready.

**Review by smcintyre-r7 — COMMENTED** [review]
This looks sane to me, I can't quite reproduce the original issue though even after running the tests 10+ times. It looks like this will just result in `core_channel_read` requests returning data from stderr in addition to stdout where as before stderr would only be read if the channel was interactive and had data to be sent to Metasploit.

Pretty sure I tracked down the cause of the `error: [Errno 32] Broken pipe` errors as well. That's unrelated and I'll submit a separate fix for that.

File: python/meterpreter/meterpreter.py [hunk 1]
```diff
@@ -680,10 +680,13 @@ def is_alive(self):
...
     def read(self, length):
-        data = ''
+        data = bytes()
         stdout_reader = self.proc_h.stdout_reader
+        stderr_reader = self.proc_h.stderr_reader
         if stdout_reader.is_read_ready():
-            data = stdout_reader.read(length)
+            data += stdout_reader.read(length)
+        if stderr_reader.is_read_ready():
+            data += stderr_reader.read(length)
```
Comment (smcintyre-r7) [inline python/meterpreter/meterpreter.py:689]: Minor suggestion to read stderr before stdout like it is in the loop around L1320 and to still keep track of the length.

```suggestion
        if stderr_reader.is_read_ready() and length > 0:
            data += stderr_reader.read(length)
        if stdout_reader.is_read_ready() and (length - len(data)) > 0:
            data += stdout_reader.read(length - len(data))
```

**Comment by timwr** [comment]
Thanks @smcintyre-r7 
I've rebased this on top of https://github.com/rapid7/metasploit-payloads/pull/514 and applied the suggestion.
Agreed the intermittent test fail is pretty hard to reproduce, also it's not a regression. I'm happy to bump the payloads repository in framework after this is landed.

**Comment by smcintyre-r7** [comment]
Looks good to me, I can see the `post/test/cmd_exec` tests are passing now.

```
msf6 post(test/cmd_exec) > show options 

Module options (post/test/cmd_exec):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION                   yes       The session to run this module on

msf6 post(test/cmd_exec) > set SESSION -1 
SESSION => -1
msf6 post(test/cmd_exec) > run

[*] Running against session -1
[*] Session type is meterpreter and platform is linux
[*] Starting cmd_exec tests
[+] should return the result of echo
[+] should return the full response after sleeping
[+] should return the full response after sleeping
[+] should return the result of echo 10 times
[*] Finished cmd_exec tests
[*] Starting cmd_exec quote tests
[+] should return the result of echo with single quotes
[+] should return the result of echo with double quotes
[*] Starting cmd_exec stderr tests
[+] should return the stderr output
[*] Testing complete in 2.829003882
[*] Passed: 7; Failed: 0
[*] Post module execution completed
```

**Pull request merged**